### PR TITLE
doc: do not document user.meta-data key

### PR DESCRIPTION
**Pull request by holmanb**
## Proposed Commit Message
```
doc: do not document user.meta-data key

This key is not documented by lxd and will likely be removed.
Also prefer cloud-init prefixes.
```

## Additional Context
https://github.com/canonical/lxd/issues/13853#issuecomment-2374472140

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

**Comment by blackboxsw** [comment]
@holmanb note that dropping this feature will affect our integration tests which use [`user.meta-data` config key to setup ssh keys for ssh](https://github.com/canonical/pycloudlib/blob/main/pycloudlib/lxd/cloud.py#L161-L165)

**Comment by TheRealFalcon** [comment]
> dropping this feature will affect our integration tests which use [user.meta-data config key to setup ssh keys for ssh](https://github.com/canonical/pycloudlib/blob/main/pycloudlib/lxd/cloud.py#L161-L165)

I think this is fairly significant. Without keys in metadata, we have to resort to using `lxc exec` everywhere, hacking the user data to contain keys, or updating the instance after lxc init has been called, but at that point the user does not exist  yet. Those all seem like pretty bad alternatives to me.

File: doc/rtd/reference/datasources/lxd.rst [hunk 1]
```diff
@@ -37,26 +36,16 @@ launched container, use either LXD profiles or
...
-- ``user.meta-data``: YAML metadata which will be appended to base metadata.
-- ``user.vendor-data``: YAML which overrides any metadata values.
-- ``user.network-config``: YAML representing either :ref:`network_config_v1` or
+- ``cloud-init.vendor-data``: YAML which overrides any metadata values.
+- ``cloud-init.network-config``: YAML representing either :ref:`network_config_v1` or
   :ref:`network_config_v2` format.
-- ``user.user-data``: YAML which takes precedence and overrides both metadata
+- ``cloud-init.user-data``: YAML which takes precedence and overrides both metadata
   and vendor data values.
 - ``user.any-key``: Custom user configuration key and value pairs, which can be
```
Comment (holmanb) [inline doc/rtd/reference/datasources/lxd.rst:44]: I was unable to find anything that corroborates this claim. I couldn't find any mention of `user.any-kay` it in the code, nor do I see the lxd datasource storing arbitrary keys anywhere that could pass through.

I tested the `cloud-init query` portion of this by setting `user.any-key` to `key: value` and cleaning the instance then rebooting. I didn't test the instance-data template portion, but I don't have high hopes.

Think we should remove it @TheRealFalcon @blackboxsw?

**Comment by holmanb** [comment]
> > dropping this feature will affect our integration tests which use [user.meta-data config key to setup ssh keys for ssh](https://github.com/canonical/pycloudlib/blob/main/pycloudlib/lxd/cloud.py#L161-L165)
> 
> I think this is fairly significant. Without keys in metadata, we have to resort to using `lxc exec` everywhere, hacking the user data to contain keys, or updating the instance after lxc init has been called, but at that point the user does not exist yet. Those all seem like pretty bad alternatives to me.

@TheRealFalcon @blackboxsw we're engaged with lxd currently in the linked bug. Would you rather wait on this docs change until we have further resolution there?

Also, I made a comment about a documented key on this page that doesn't appear to align with cloud-init's behavior.

**Comment by blackboxsw** [comment]
@holmanb I'm leaning toward moving forward with this change. We are documenting things that are not even documented  best practices in LXD currently, So I think it's reasonable for us to avoid such documentation in cloud-init, it will be stale at best.

**Comment by TheRealFalcon** [comment]
Was about to say something similar, but @blackboxsw beat me to it

**Review by blackboxsw — APPROVED** [review]
LGTM on the doc changes here. Dropping `user.meta-data` from docs makes sense regardless of whether LXD drops this, or changes it to a different feature as LXD has dropped this config option from their docs too. We will only be stale at best, and broken if that config options drops from LXD completely in the future.

What remains is whether we should document the potential uses of `user.any-key` with an example to query custom user.my-key values.  (maybe in our lxd tutorial?).  I don't think that needs to block this PR, but it's something we can discuss or provide separately if we feel it's needed.

File: doc/rtd/reference/datasources/lxd.rst (hunk 1 above)
Comment (holmanb) [reply doc/rtd/reference/datasources/lxd.rst:44]: > ```
> $ ': lxc launch ubuntu-daily:jammy -c user.my-tag=neato test-custom-user-config
> $ lxc exec test-custom-user-config -- cloud-init query ds.config.user_my_tag
> neato
> # jinja template custom behavior
> $ lxc exec test-custom-user-config -- cloud-init query --format "{% if ds.config.user_my_tag == 'neato'  %}sh /custom_script.sh{% endif %}"
> /custom_script.sh
> ```
> 
> Key aliases are provided by cloud-init for ease of use in jinja templates and queries
> 
> ```
> $ cloud-init query --format "{{ds.config}}"
> {'user.my-key': 'two', 'user_my_key': 'two'}
> ```

Thanks for the clarification. Reading this doc, I interpreted it as saying that key `user.any-key` is literal, not arbitrary, and expected key pairs set in `user.any-key` to be the key/values that would be exposed / visible in `cloud-init query` and in template/jinja. I see now that `user.any-key` _is_ a key. I'll modify it to hopefully avoid others from interpreting it like I did.

**Review by blackboxsw — APPROVED** [review]
LGTM, ship it!

File: doc/rtd/reference/datasources/lxd.rst [hunk 2]
```diff
@@ -37,25 +36,28 @@ launched container, use either LXD profiles or
...
-  which can be used by both `#template: jinja` #cloud-config templates and
-  the :command:`cloud-init query` command.
+- ``cloud-init.vendor-data``: YAML which overrides any metadata values.
+- ``cloud-init.network-config``: YAML representing either
+  :ref:`network_config_v1` or :ref:`network_config_v2` format.
+- ``cloud-init.user-data``: YAML which takes precedence and overrides both
+  metadata and vendor data values.
+- ``user.<any-key>``: Keys prefixed with ``user.`` are included in
+  :ref:`instance data<instance_metadata>`. under the ``ds.config`` key. These
+  key-value pairs are used in jinja-templated :ref:`cloud-config<jinja-config>`
```
Comment (blackboxsw) [inline doc/rtd/reference/datasources/lxd.rst:46]: docs CI test doesn't like this line:
`doc/rtd/reference/datasources/lxd.rst:45: : Spell check: templated: key-value pairs are used in jinja-templated .`

**Pull request merged**